### PR TITLE
chore(agw): Declare conflicting packages

### LINKED ISSUE #14709: Make package conflicts explicit
For the Magma .deb package built without Bazel, we build a separate .deb package for each Python dependency. The Magma package has all those Python .deb packages listed as dependencies.
The bazel-built .deb Package contains all its Python dependencies.

Tthe bazel-built .deb package puts the files in the same location as the old Python .deb packages. This makes the packages conflict, so that a Magma upgrade requires manual conflict resolution. A longterm solution might be to have the new self-contained package put all files to some isolated location so that we don't get conflicts with existing packages in the first place. This ticket instead focuses on a quickfix: **Declare the conflicts in the package metadata, so that `apt` is able to resolve the conflicts automatically, thus making the upgrade more seamless**.

### PR DESCRIPTION
**Pull request by jheidbrink**
## Summary

#closes #14709

Declare `replaces` and `provides` fields for the Magma package

## Test Plan

I encountered difficulties passing a .deb filename as parameter to `apt install` - `apt` would usually prefer to install a package of the same name from its configured repositories. Thus I used the following script to create ad-hoc repositories containing the exact packages I want to test:  (if you want to try the script, copy a 1.8 package and a 1.9 package to your working directory and adapt the filenames in the script):
```
#!/usr/bin/env bash

set -euo pipefail

#aptly mirror create -ignore-signatures -filter='"!Name (magma)" | "!Name(magma-sctpd)"' thirdparty https://linuxfoundation.jfrog.io/artifactory/magma-packages-test focal-ci
# the above filter doesn't work. the command below does not exclude magma-sctpd and thus downloads 11GB of packages which is annoying but works
aptly mirror create -ignore-signatures -filter='!Name (magma)' thirdparty https://linuxfoundation.jfrog.io/artifactory/magma-packages-test focal-ci
aptly mirror update -ignore-signatures thirdparty
aptly snapshot create thirdparty-snapshot from mirror thirdparty
aptly publish --skip-signing -architectures=amd64 snapshot thirdparty-snapshot
# clean up via
#aptly publish drop focal-ci
#aptly snapshot drop thirdparty-snapshot
#aptly mirror drop thirdparty

aptly repo create magma18
aptly repo add magma18 magma_1.8.0-1663094999-49ed2e69_amd64.deb
aptly publish --skip-signing -distribution=magma18 -architectures=amd64 repo magma18

aptly repo create magma19
aptly repo add magma19 magma_1.9.0-1671632057-e9512c02_amd64.deb
aptly publish --skip-signing -distribution=magma19 -architectures=amd64 repo magma19
# clean up via
#aptly publish drop magma19
#aptly repo drop magma19

aptly serve &

docker run --network=host --interactive ubuntu:focal <<EOT
echo 'deb [trusted=yes] http://localhost:8080/ focal-ci main' > /etc/apt/sources.list.d/focal-ci-http.list
echo 'deb [trusted=yes] http://localhost:8080/ magma18 main' > /etc/apt/sources.list.d/magma18-http.list
apt update
env DEBIAN_FRONTEND=noninteractive apt install -y systemd psmisc  # postinst script dependencies (note that they don't have set -e, so they dont fail)
env DEBIAN_FRONTEND=noninteractive apt install -y magma

echo 'deb [trusted=yes] http://localhost:8080/ magma19 main' > /etc/apt/sources.list.d/magma19-http.list
apt update
env DEBIAN_FRONTEND=noninteractive apt install -y magma
EOT
```

## Additional information

Besides building and uploading custom Python packages for third party dependencies, we also build and upload our own .deb packages for Python packages that are already packaged in Ubuntu. Our custom .deb packages go by the same names as the upstream .deb packages. They contain the same files, but in different locations.

For determining conflicting packages that go by the same name as upstream packages, I could use the script below. For thirdparty packages this approach didn't work and I used a trial-and-error approach: try to install the package, see it fail, add the conflict metadata, try again.

The script below takes a long time to run and produces a lot of output. Conflicting packages are printed to stdout, and log output is printed to stderr.

```
#!/usr/bin/env bash

set -euo pipefail

function log() {
    echo "$@" >&2
}

function list_magma_files() {
  bazel run //lte/gateway/release:release_build > /dev/null
  dpkg --contents $(ls -1 -t /tmp/packages/magma_1.9* | head -n 1)
}

function no_directories() {
  sed -e '/.*\/$/d'  # if it ends with a slash, it's a directory
}

function find_package() {
    filename="$1"
    log "Looking for package for $filename"
    if dpkg -S "$filename" > /dev/null 2>/dev/null
    then
        log "Found via dpkg"
        dpkg -S "$filename" | cut -d ':' -f 1
    else
        adapted=$(echo "$filename" | sed -e 's|/usr/local/|/usr/|' -e 's|python3.8|python3|')
        if apt-file find "$adapted" > /dev/null 2>/dev/null
        then
          log "Found via apt-file"
          apt-file find "$adapted" | cut -d ':' -f 1
        else
          log "No package found for $filename"
        fi
    fi
}

function find_packages() {
    while read -r filename
    do
      find_package "/${filename}"
    done
}

list_magma_files | awk '{print $6}' | no_directories | find_packages | uniq
```

### CODE REVIEW

File: lte/gateway/release/BUILD.bazel [hunk 1]
```diff
@@ -229,8 +229,14 @@ pkg_deb(
     package = MAGMA_PKGNAME,
     package_file_name = "{fname}.deb".format(fname = MAGMA_FILE_NAME),
     postinst = ":magma-postinst-bazel",
-    provides = [MAGMA_PKGNAME],
```
Comment (jheidbrink) [inline lte/gateway/release/BUILD.bazel:246]: I think the old `provides` and `replaces` fields were unnecessary. While it's possible to set it, it seems this is not done by upstream Ubuntu packages.

Comment (sebathomas) [reply lte/gateway/release/BUILD.bazel:246]: I think it's done by fpm (I can see the fields for magma-1.8.0) and we probably reproduced it in Bazel to stay consistent. But I agree it doesn't seem to be necessary.

File: lte/gateway/release/deb_dependencies.bzl [hunk 2]
```diff
@@ -78,3 +78,125 @@ OVS_DEPS = [
...
+    "python3-tz",
+    "python3-urllib3",
+    "python3-webcolors",
+    "python3-webob",
+    "python3-websocket",
+    "python3-werkzeug",
+    "python3-wrapt",
+    "python3-yaml",
+    "python3-yarl",
+]
```
Comment (sebathomas) [inline lte/gateway/release/deb_dependencies.bzl:174]: I can't really think of a way to check that these are consistent with the Python dependencies of Magma. Maybe we could add a comment here and in the requirements.in?

**Review by sebathomas — COMMENTED** [review]
Definitely out of scope for this ticket: My favorite solution would be to install all Python dependencies of Magma to some different location like a virtualenv, and to make sure that magma uses these dependencies. Then we would be independent from other Python packages installed by the OS.

But I think the approach in this PR is a good improvement for the current solution.

**Comment by sebathomas** [comment]
There seems to be a Bazel formatting issue, the CI is red.

File: lte/gateway/release/BUILD.bazel (hunk 1 above)
Comment (nstng) [reply lte/gateway/release/BUILD.bazel:246]: confirm: I took this over to the bazel build for consistency - if it's unnecessary, then we can remove it.

**Comment by nstng** [comment]
Reproduced and verified for review:

# Reproduce

```
docker run --network=host -it ubuntu:focal /bin/bash
apt install gnupg wget
wget -qO - https://linuxfoundation.jfrog.io/artifactory/api/security/keypair/magmaci/public | apt-key add -
echo 'deb https://linuxfoundation.jfrog.io/artifactory/magma-packages-test focal-ci main' > /etc/apt/sources.list.d/focal-ci-http.list
apt update
DEBIAN_FRONTEND=noninteractive apt install magma=1.8*
DEBIAN_FRONTEND=noninteractive apt install magma

dpkg: error processing archive /var/cache/apt/archives/magma_1.9.0-1671702983-74e4a1b3_amd64.deb (--unpack):
 trying to overwrite '/usr/local/lib/python3.8/dist-packages/Crypto/Cipher/AES.py', which is also in package python3-pycryptodome 3.15.0
```

# Verify

```
# devcontainer
bazel run //lte/gateway/release:release_build --config=production
cp /tmp/packages/magma* .

# host in magma root
aptly repo create magma19
aptly repo add magma19 magma*
aptly publish --skip-signing -distribution=magma19 -architectures=amd64 repo magma19
aptly serve

docker run --network=host -it ubuntu:focal /bin/bash
apt update
apt install gnupg wget
wget -qO - https://linuxfoundation.jfrog.io/artifactory/api/security/keypair/magmaci/public | apt-key add -
echo 'deb https://linuxfoundation.jfrog.io/artifactory/magma-packages-test focal-ci main' > /etc/apt/sources.list.d/focal-ci-http.list
apt update
DEBIAN_FRONTEND=noninteractive apt install magma=1.8*
echo 'deb [trusted=yes] http://localhost:8080/ magma19 main' > /etc/apt/sources.list.d/magma19-http.list
apt update
DEBIAN_FRONTEND=noninteractive apt install magma
echo $?
0
```

**Review by nstng — APPROVED** [review]

File: lte/gateway/release/deb_dependencies.bzl (hunk 2 above)
Comment (jheidbrink) [reply lte/gateway/release/deb_dependencies.bzl:174]: I think they don't need to be kept in sync. Assuming that with Magma 1.9 we switch to the bazel-built .deb package which includes all dependencies, the problem this PR addresses is only for upgrades from Magma < 1.9 to Magma 1.9. Future upgrades should be fine. And assuming that we won't change the requirements for Magma versions < 1.9 anymore, it should be fine for requirements.txt and this list here to diverge on master.

**Review by sebathomas — APPROVED** [review]

**Comment by sebathomas** [comment]
For some reason the CI seems to be stuck.

**Comment by jheidbrink** [comment]
> Definitely out of scope for this ticket: My favorite solution would be to install all Python dependencies of Magma to some different location like a virtualenv, and to make sure that magma uses these dependencies. Then we would be independent from other Python packages installed by the OS.

I absolutely agree and created https://github.com/magma/magma/issues/14746

**Pull request merged**